logshifter/input_test.go

```diff
@@ -0,0 +1,90 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+	"sync"
+	"testing"
+)
+
+// TestInputBlocking ensures that Input never blocks, regardless of the queue
+// consumption rate.
+func TestInputBlocking(t *testing.T) {
+	messageCount := int64(10000)
+	messageSize := 1024
+
+	// Use a small queue to help ensure an input backup.
+	queue := make(chan []byte, 1)
+
+	// Set up a fast output sink to help ensure an input backup.
+	go func() {
+		for {
+			<-queue
+		}
+	}()
+
+	// Collect stats.
+	stats := make(map[string]float64)
+	statsCh := make(chan Stat)
+	stop := make(chan struct{})
+	statsWg := sync.WaitGroup{}
+	statsWg.Add(1)
+	go func() {
+		for {
+			select {
+			case <-stop:
+				statsWg.Done()
+				return
+			case s := <-statsCh:
+				stats[s.name] = stats[s.name] + s.value
+			}
+		}
+	}()
+
+	// Read until the test reader closes.
+	reader := newTestReader(messageCount, messageSize)
+	input := &Input{
+		reader:       reader,
+		queue:        queue,
+		bufferSize:   messageSize,
+		statsChannel: statsCh,
+	}
+	input.read()
+
+	// Shut down the stats collector.
+	close(stop)
+	statsWg.Wait()
+	t.Logf("stats: %#v", stats)
+}
+
+// testReader is an io.Reader which will read from a predefined buffer and
+// returns an error when the buffer is empty.
+type testReader struct {
+	buffer *bytes.Buffer
+}
+
+func (r *testReader) Read(b []byte) (int, error) {
+	written, err := r.buffer.Read(b)
+
+	if err != nil {
+		return written, err
+	}
+
+	return written, nil
+}
+
+func newTestReader(msgCount int64, msgLength int) *testReader {
+	var data string
+
+	var i int64 = 0
+	for ; i < msgCount; i++ {
+		data += strings.Repeat("0", msgLength-1) + "\n"
+	}
+
+	buffer := bytes.NewBufferString(data)
+
+	fmt.Printf("created %d byte test input (%d lines @ %d bytes each)\n", buffer.Len(), msgCount, msgLength)
+
+	return &testReader{buffer: buffer}
```

Comment:
You're right; I copied and gutted that structure from elsewhere, and in the gutting removed the reasons for the struct to exist. Simplified.